Improvements to CSV export

m-col opened the pull request:
### Changes

Following on from #1387, these two commits make two changes to the CSV/ZIP export feature.

> in pages.csv both the bounce_rate and time_on_page are empty
> in sources.csv both the bounce_rate and visit_duration are empty

The first commit addresses this. CSVs will not provided all detail for pages.csv, sources.csv, exits.csv and entries.csv.

> Renaming:
> 
> count -> visitors
> 
> In all reports except:
> 
> In conversions.csv:
> Count -> unique_conversions
> total_count -> total_conversions
> 
> In entry_pages.csv:
> count -> unique_entrances
> entries -> total_entrances
> 
> In exit_pages.csv:
> count -> unique_exits
> exits -> total_exits

The second commit makes these changes, renaming some of the data as provided via the API. This applies to both the JSON and CSV results, which provide the same labels for each metric.

Note the two other changes that are not part of this PR are 1. improve download UX and 2. output full country names in countries.csv. I can append those to this PR when implemented, or post as a follow-up PR, whichever is preferable. EDIT: the country names have been added to this PR, as has fetching data concurrently.

### Tests
- [x] Automated tests have been ~~added~~ updated

### Changelog
- [x] Entry has been added to changelog

### Documentation
- [x] This change does not need a documentation update

m-col commented on lib/plausible_web/controllers/stats_controller.ex:
```diff
@@ -64,8 +64,8 @@ defmodule PlausibleWeb.StatsController do
-    params = Map.put(params, "limit", "1000")
-    params = Map.put(params, "csv", "True")
+    # If the limit is set to 363 or more, ClickHouseRepo.query fails when fetching breakdown
+    params = Map.merge(params, %{"limit" => "362", "csv" => "True", "detailed" => "True"})
```
This is here as a bit of a temporary note - I'd like to track down the reason why it's necessary so we don't need to guess a number

m-col replied:
For more info, what seems to trigger the error is using a large limit (e.g. 1000) and then requesting only a small amount of data (e.g. last 7 days with the test data I am using). In theory 300 may still cause problems if filtering for less data and/or using a smaller dataset. The error is:

```
Request: GET /plausible.io/export?period=7d&date=2021-11-01&filters=%7B%7D
** (exit) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, %Clickhousex.Error{code: "42", constraint_violations: [], message: "Code: 42, e.displayText() = DB::Exception: Function tuple requires at least one argument.: While processing tuple(): While processing (s = s2) AND (p IN tuple()) (version 21.4.6.55 (official build))\n"}}
        (plausible 0.0.1) lib/plausible/stats/breakdown.ex:246: Plausible.Stats.Breakdown.breakdown_time_on_page/3
        (plausible 0.0.1) lib/plausible/stats/breakdown.ex:100: Plausible.Stats.Breakdown.breakdown/5
```

ukutaht replied:
Oh, OK. I thought it was some more obscure error. This looks like it can be tracked down and fixed

ukutaht replied:
If you have unfinished stuff like this in the PR, you can mark the PR as a draft in Github.

m-col replied:
Good idea

m-col commented on assets/js/dashboard/stats/bar.js:
```diff
@@ -1,17 +1,20 @@
-  const width = barWidth(count, all)
+export default function Bar({count, all, bg, maxWidthDeduction, children, plot = "visitors"}) {
+	if (plot != "visitors") {
+		console.log(all);
```
Ah yep

m-col commented:
Fixed up those two comments, and I've added a further commit to fix the time being used for adding test data for one of the tests, which was causing failures on master.

m-col commented:
Hmm, I am getting that error again even when there are pages available so further investigation is needed.

m-col commented:
The error I am seeing now is the one I mentioned last week, that was more obscure than the one mentioned earlier. The trace is:
```
Request: GET /plausible.io/export?period=6mo&date=2021-11-01&filters=%7B%7D
** (exit) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, %Clickhousex.Error{code: :unknown, constraint_violations: [], message: ""}}
        (plausible 0.0.1) lib/plausible/stats/breakdown.ex:250: Plausible.Stats.Breakdown.breakdown_time_on_page/3
        (plausible 0.0.1) lib/plausible/stats/breakdown.ex:100: Plausible.Stats.Breakdown.breakdown/5
```
The query call itself, from `lib/plausible/stats/breakdown.ex`, is:
```
{:ok, res} = ClickhouseRepo.query(time_query, base_query_raw_params ++ [pages])
```
where pages contains (in this case) 617 items, and the limit is 1000. This was the case where I found setting the limit to a specific number caused the query to error, which now appears to be 364. The constraint violation and message are not especially useful here... any ideas as to where to investigate this?

m-col commented:
I've added some more things that we discussed:

1. Run database requests in parallel
2. Display a  notification when the download has been requested. The notification uses the same html+css as the `put_flash`-triggered notifications, but is triggered on the JS side when the button is clicked. This allows for the page to stay visible rather than refreshing the page, without the need for setting up a whole Channel setup in the pheonix app.
3. Fade and disable the download  button for the 4s while the notification is visible.

I think that was everything, so the lat thing is just to figure out that bug. I haven't managed to track it down unfortunately.

ukutaht commented on assets/js/dashboard/stats/visitor-graph.js:
```diff
@@ -301,12 +301,20 @@ class LineGraph extends React.Component {
     return stats
   }
 
+  downloadHandler() {
```
This solution mixes React with manual DOM manipulation and Elixir rendering which is not a good idea. The dashboard is rendered in React and any subsequent UI interactions should not leave React.

Manual DOM manipulation can mess up React's virtual DOM and lead to runtime errors. Plus, it's just confusing to switch mental models as you read the code or try to debug it.

The reason why I preferred a spinner to a notification in our last call is because we don't have a notification system in React. It requires some slightly advanced knowledge of React to accomplish that.

I would suggest just replacing the icon with a spinner which is much easier to do without leaving React. Alternatively, you can look into how to build notifications with React (without bringing in extra dependencies like Redux)

m-col commented:
Unfortunately I have had no luck tracking down the clickhouse query error. The error is:
```
Clickhousex.Error{code: :unknown, constraint_violations: [], message: ""}
```
which is a pretty useless error. My last theory was that the get request sent to the clickhouse server was too long but printing the length of the request ruled that out, and in fact when I set the limit to `363` (number may vary depending on data being requested) it _sometimes_ successeded and _sometimes_ triggered this error. I'm not seeing anything useful printed, beside potentially some `DBConnection.ConnectionError)` errors which may be related. However looking this up does not point to any particular error.

Anyway, the 300 limit might be enough in practice but could be increased if there is demand for that. If so this error may need further investigation.

A spinner has replaced the notification window so otherwise this PR is good to go.

metmarkosaric commented:
just tested the full csv export - few more comments:

- spinner works/looks great!
- in conversions.csv the unique_conversions is empty
- in pages.csv time_on_page entries have decimal points (example: 1219.28019323672) - could it be like visitor duration so it's displayed without decimals?
- on last 12 months the download doesn't start when spinner stops (last 6 months works fine)... i never got last 12 months to work but at least no sentry errors any more now

everything else seems fine including all things we discussed from the last round of testing! we're getting there!
